### PR TITLE
AutoMatedTest support test module.parameter.grad

### PR DESCRIPTION
**Pull request by wyg1997**
- 增强 AutoMatedTest 的功能，使之比较 module.parameter.grad 的值
- 修复 bn 层 grad 计算不对齐的问题

这一修改可能会暴露出更多 module 后向实现有误的问题，需要逐一修改

### CODE REVIEW

File: python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py [hunk 1]
```diff
@@ -410,8 +410,15 @@ def new_f(test_case):
...
+                                getattr(x.oneflow, key),
+                            )
+                        )
+                        dual_objects_to_test.append(
+                            GetDualObject(
+                                "unused",
+                                getattr(x.pytorch, key).grad,
+                                getattr(x.oneflow, key).grad,
                             )
                         )
```
Comment (wyg1997) [inline python/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py:423]: 增加参数梯度的对比

File: python/oneflow/test/modules/test_batchnorm.py [hunk 2]
```diff
@@ -27,6 +27,20 @@
 
 @flow.unittest.skip_unless_1n1d()
 class TestBatchNormModule(flow.unittest.TestCase):
+    @autotest(n=20, auto_backward=True, rtol=1e-3, atol=1e-3)
+    def test_batchnorm1d_module_with_random_data(test_case):
+        device = random_device()
+        channel = random(1, 4).to(int)
+        m = torch.nn.BatchNorm1d(num_features=channel, track_running_stats=False).to(
```
Comment (hjchen2) [inline python/oneflow/test/modules/test_batchnorm.py:34]: 是不是可以把`track_running_stats =True`的单测也加上了

**Comment by wyg1997** [comment]
目前 bn 的 cpu 实现还没有对齐，torch 的计算公式为：

```
  /// Collect the linear and constant terms regarding the input.
  /// output(n, c, h, w)
  ///     = (input(n, c, h, w) - mean(c)) / sqrt(var(c) + eps) * weight(c)
  ///         + bias(c)
  ///     = input(n, c, h, w) * inv_var(c) * weight(c)
  ///         - mean(c) * inv_var(c) * weight(c) + bias(c),
  /// where inv_var(c) = 1 / sqrt(var(c) + eps).
  /// So the linear term, alpha(c) = inv_var(c) * weight(c),
  ///   the constant term beta(c) = bias(c) - mean(c) * inv_var(c) * weight(c)
  /// Note that this is only a good idea if (input_size >> c), in degenerate
  /// cases where image_size == 1 && batch_size == 1, it is slow.
```

**Comment by hjchen2** [comment]
> 目前 bn 的 cpu 实现还没有对齐，torch 的计算公式为：
> 
> ```
>   /// Collect the linear and constant terms regarding the input.
>   /// output(n, c, h, w)
>   ///     = (input(n, c, h, w) - mean(c)) / sqrt(var(c) + eps) * weight(c)
>   ///         + bias(c)
>   ///     = input(n, c, h, w) * inv_var(c) * weight(c)
>   ///         - mean(c) * inv_var(c) * weight(c) + bias(c),
>   /// where inv_var(c) = 1 / sqrt(var(c) + eps).
>   /// So the linear term, alpha(c) = inv_var(c) * weight(c),
>   ///   the constant term beta(c) = bias(c) - mean(c) * inv_var(c) * weight(c)
>   /// Note that this is only a good idea if (input_size >> c), in degenerate
>   /// cases where image_size == 1 && batch_size == 1, it is slow.
> ```

我们cpu和这个计算公式应该就是对齐的吧，后面那个等式可能会造成精度差异。

**Comment by wyg1997** [comment]
> 我们cpu和这个计算公式应该就是对齐的吧，后面那个等式可能会造成精度差异。

我又核对了一下，是 running_mean 和 running_var 错了

File: python/oneflow/nn/modules/batchnorm.py [hunk 3]
```diff
@@ -114,14 +119,15 @@ def forward(self, x):
...
-                running_mean = (
-                    self.momentum * self.running_mean + (1 - self.momentum) * mean
-                )
-                running_var = (
-                    self.momentum * self.running_var + (1 - self.momentum) * variance
-                )
-                self.__setattr__("running_mean", running_mean)
-                self.__setattr__("running_var", running_var)
+                # use unbiased variance to update running_var
+                unbiased_variance = x.var(dim=reduce_axis, unbiased=True, keepdim=False)
```
Comment (wyg1997) [inline python/oneflow/nn/modules/batchnorm.py:123]: 更新 running_var 时用了无偏估计，后面计算的时候用的是真正的方差

**Review by hjchen2 — APPROVED** [review]

**Review by hjchen2 — APPROVED** [review]

**Review by BBuf — APPROVED** [review]

**Comment by BBuf** [comment]
自动测试给出哪个 module 的哪个参数没对齐的提示，这个是打算怎么做呢？

**Comment by wyg1997** [comment]
> 自动测试给出哪个 module 的哪个参数没对齐的提示，这个是打算怎么做呢？

游离的 tensor 不好办，module 里的参数都带名字的，这个在创建比较集合的时候就把名字传进去，对比出错打印的时候可以打出来
